### PR TITLE
feat: data validation handler

### LINKED ISSUE #6: Handle data validation requests
- `POST /v0/data-deal/deals/{dealId}/data`
  - Read a data from request body
  - Decrypt data
  - Validate data
  - Re-encrypt data using a combined key
  - Put data into IPFS
  - Issue a certificate to the client

### PR DESCRIPTION
**Pull request by H4NLee**
close #6

### CODE REVIEW

**Comment by H4NLee** [comment]
I have not tested full cycle of this handler because some relevant functions of Panacea are in developing.
After implementation of those, I'll test it.

**Comment by H4NLee** [comment]
I can make test codes for this handler as youngjoon did for `auth_test.go`.
Will add soon

File: service/handler.go [hunk 1]
```diff
@@ -1,8 +1,175 @@
...
+	"github.com/medibloc/panacea-oracle/crypto"
+	"github.com/medibloc/panacea-oracle/validation"
+	log "github.com/sirupsen/logrus"
+)
+
+type ValidateDataReq struct {
+	ProviderAddress     string `json:"provider_address"`
+	EncryptedDataBase64 string `json:"encrypted_data_base64"`
+	DataHash            string `json:"data_hash"`
+}
```
Comment (gyuguen) [inline service/handler.go:25]: How about added validation function like `func (m ValidateDataReq) ValidateBasic() {...}`.
This is a validation of request parameters.

File: service/handler.go [hunk 2]
```diff
@@ -1,8 +1,175 @@
...
+		return
+	}
+
+	deal, err := queryClient.GetDeal(dealID)
+	if err != nil {
+		log.Errorf("failed to get deal(%d): %s", dealID, err.Error())
+		http.Error(w, "failed to get deal", http.StatusBadRequest)
+		return
+	}
+
```
Comment (gyuguen) [inline service/handler.go:102]: How about added validation a deal status?

File: service/handler.go [hunk 3]
```diff
@@ -1,8 +1,175 @@
...
+	// Put data into IPFS
+	cid, err := svc.IPFS().Add(reEncryptedData)
+	if err != nil {
+		log.Errorf("failed to store data to IPFS: %s", err.Error())
+		http.Error(w, "failed to store data to IPFS", http.StatusInternalServerError)
+		return
+	}
+
+	// Issue a certificate to the client
+	unsignedDataCert := &datadealtypes.UnsignedCertificate{
```
Comment (gyuguen) [inline service/handler.go:127]: I think we need to the `uniqueID` parameter.
However, it is not supported by the current panacea-core version.

You can add a unique ID later. :)

**Review by gyuguen — APPROVED** [review]
Thanks.
I left a few comments.

File: service/handler.go (hunk 3 above)
Comment (H4NLee) [reply service/handler.go:127]: Right. I'll add it after some PRs in panacea-core are merged

**Review by youngjoon-lee — COMMENTED** [review]
I'm approving merging this PR in general, but it seems there are some duplicated codes with PR #21, in regard to the AES encryption.

**Review by youngjoon-lee — APPROVED** [review]

**Review by audtlr24 — APPROVED** [review]
Thanks, lgtm

**Comment by H4NLee** [comment]
I tried to write test for this handler, but it requires to mock service, and many components inside service (query client, ipfs, etc.).
So I decided not to write test for this handler using mocking. I think we need to test this in e2e test.

**Comment by youngjoon-lee** [comment]
> I tried to write test for this handler, but it requires to mock service, and many components inside service (query client, ipfs, etc.).
> So I decided not to write test for this handler using mocking. I think we need to test this in e2e test.

Yup. No worries :)

File: crypto/aes256.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"fmt"
+	"io"
+)
+
+// Encrypt combines secretKey and secondKey to encrypt with AES256-GCM method.
+func Encrypt(secretKey, additional, data []byte) ([]byte, error) {
```
Comment (inchori) [inline crypto/aes256.go:12]: Ah, If these functions are merged into main branch, I'll change it when executing `approveOracleRegistration` tx.

**Review by inchori — APPROVED** [review]
LGTM!

**Pull request merged**